instance/tests/management/test_reprovision_buckets.py

```diff
@@ -0,0 +1,87 @@
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance - reprovision_buckets unit tests
+"""
+from unittest.mock import patch
+
+from django.core.exceptions import FieldDoesNotExist
+from django.core.management import call_command
+from django.test import TestCase
+from django.utils.six import StringIO
+from testfixtures import LogCapture
+
+from instance.models.openedx_instance import OpenEdXInstance
+
+
+class ReprovisionBucketsTestCase(TestCase):
+    """
+    Test cases for the `reprovision_buckets` management command.
+    """
+
+    def test_no_instances(self):
+        """
+        Verify that the command correctly notifies the user that there are no instances for migration.
+        """
+        with LogCapture() as captured_logs:
+            call_command(
+                'reprovision_buckets',
+                stdout=StringIO(),
+            )
+            # Verify the logs
+            self.assertIn(
+                'Found "0" active instances',
+                [l[2] for l in captured_logs.actual()])
+
+    @patch('instance.models.mixins.storage.S3BucketInstanceMixin._enable_bucket_versioning')
+    @patch('instance.models.mixins.storage.S3BucketInstanceMixin._update_bucket_lifecycle')
+    @patch('instance.models.mixins.storage.S3BucketInstanceMixin._update_bucket_cors')
+    @patch('instance.models.mixins.storage.S3BucketInstanceMixin._perform_create_bucket')
+    @patch('instance.models.mixins.storage.S3BucketInstanceMixin._update_iam_policy')
+    @patch('instance.models.mixins.storage.S3BucketInstanceMixin.create_iam_user')
+    def test_migrate(self,
+                     mock_create_iam_user,
+                     mock_update_iam,
+                     mock_create_bucket,
+                     mock_update_cors,
+                     mock_update_lifecycle,
+                     mock_enable_versioning):
+        """
+        Verify that the command correctly reprovision the bucket for an instance.
+        """
+        OpenEdXInstance.objects.create(
+            sub_domain='test', name='test instance', storage_type=OpenEdXInstance.S3_STORAGE)
+        with LogCapture() as captured_logs:
+            call_command(
+                'reprovision_buckets',
+                stdout=StringIO(),
+            )
+            # Verify the logs
+            actual = set(l[2] for l in captured_logs.actual())
+            expected = {
+                'Found "1" active instances',
+            }
+            self.assertTrue(expected <= actual)
+
+        mock_create_iam_user.assert_called_once_with()
+        mock_update_iam.assert_called_once_with()
+        mock_create_bucket.assert_called_once_with('')
+        mock_update_cors.assert_called_once_with()
+        mock_update_lifecycle.assert_called_once_with()
+        mock_enable_versioning.assert_called_once_with()
```

Comment:
While these tests look technically correct, I'm not sure about the utility of them. It's more of a philosophical argument, really. It seems to be testing the implementation rather than the intended behaviour. I think elements of the underlying calls are tested elsewhere, perhaps, but really what I'd like to see here is that calling the management command results in a call to `StorageContainer._s3_client.put_bucket_lifecycle_configuration()`, `StorageContainer._iam_client.put_user_policy()`, etc.

This is a difficult one, because those clients aren't our own code (at least I don't think they are), and we'd have to mock their public interfaces - but those interfaces shouldn't change unless we change which implementation of the client we use, e.g. by upgrading it, in which case if the public API has changed, we *want* the test to fail.

Reply:
During the migration to `boto3` the testing of the actual calls to the API are being tested by stubbing.

The idea here is testing the behavior of the command, making sure it perform all the updates necessary. I was actually a bit overzealous here as there's no need to test create_iam and create_bucket.

Reply:
I'm always wary of creating brittle tests that make it difficult to refactor the implementation later - if I have to change tests to change the implementation, I tend to find I've lost some of the benefit of having the tests in the first place.

In any case, the tests do look correct, and it's hugely subjective, so I can live with this. :smile:

Reply:
I agree with that, but with  `boto3` it got out of hand given there's no possibility of mocking it. Stubbing is required and it's an order of magnitude more complex and [verbose](https://github.com/open-craft/opencraft/blob/615f99ac9ac64c308c168623de2f31873aa602b8/instance/tests/models/utils.py).

That's why I chose to encapsulate the calls to `boto3`, assuming these are correct, since these are being tested [here](https://github.com/open-craft/opencraft/pull/377/files#diff-337408cef240bc3cb85bbc4d3088bb6cR373).

Mainly, I'm checking that the management command has the appropriate behavior of updating the user policy, bucket cors, versioning and lifecycle. I have removed the unnecessary assertions.